Tests/AsyncHTTPClientTests/TransactionTests.swift

```diff
@@ -0,0 +1,585 @@
+                return XCTFail("Expected to have an HTTP2 connection here.")
+            }
+
+            let streamWriter = AsyncSequenceWriter<ByteBuffer>()
+            XCTAssertFalse(streamWriter.hasDemand, "Did not expect to have a demand at this point")
+
+            var request = HTTPClientRequest(url: "https://localhost:\(httpBin.port)/")
+            request.method = .POST
+            request.headers = ["host": "localhost:\(httpBin.port)"]
+            request.body = .stream(length: 800, streamWriter)
+
+            var maybePreparedRequest: PreparedRequest?
+            XCTAssertNoThrow(maybePreparedRequest = try PreparedRequest(request))
+            guard let preparedRequest = maybePreparedRequest else {
+                return
+            }
+            let (transaction, responseTask) = Transaction.makeWithResultTask(
+                request: preparedRequest,
+                preferredEventLoop: eventLoopGroup.next()
+            )
+
+            http2Connection.executeRequest(transaction)
+
+            XCTAssertEqual(delegate.hitStreamClosed, 0)
+
+            let response = try await responseTask.result.get()
+
+            XCTAssertEqual(response.status, .ok)
+            XCTAssertEqual(response.version, .http2)
+            XCTAssertEqual(delegate.hitStreamClosed, 0)
+
+            let iterator = SharedIterator(response.body.filter { $0.readableBytes > 0 }.makeAsyncIterator())
+
+            // at this point we can start to write to the stream and wait for the results
+
+            for i in 0..<100 {
+                let buffer = ByteBuffer(integer: i)
+                streamWriter.write(buffer)
+                var echoedBuffer = try await iterator.next()
+                guard let echoedInt = echoedBuffer?.readInteger(as: Int.self) else {
+                    XCTFail("Expected to not be finished at this point")
+                    break
+                }
+                XCTAssertEqual(i, echoedInt)
+            }
+
+            XCTAssertEqual(delegate.hitStreamClosed, 0)
+            streamWriter.end()
+            let final = try await iterator.next()
+            XCTAssertNil(final)
+            XCTAssertEqual(delegate.hitStreamClosed, 1)
+        }
+        #endif
+    }
+}
+
+#if compiler(>=5.5) && canImport(_Concurrency)
+
+// This needs a small explanation. If an iterator is a struct, it can't be used across multiple
+// tasks. Since we want to wait for things to happen in tests, we need to `async let`, which creates
+// implicit tasks. Therefore we need to wrap our iterator struct.
+@available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+actor SharedIterator<Iterator: AsyncIteratorProtocol> {
+    private var iterator: Iterator
+
+    init(_ iterator: Iterator) {
+        self.iterator = iterator
+    }
+
+    func next() async throws -> Iterator.Element? {
+        var iter = self.iterator
+        defer { self.iterator = iter }
+        return try await iter.next()
```

Comment:
I think this will only really work in the general case if Iterator has value semantics. Otherwise, concurrent access will still result in two next calls. AFAIK, our current test do not call `next()` concurrently but maybe still worth documenting or adding some precondition that checks we always leave `next()` before we enter `next()` it again.

Reply:
We have a precondition for this...
```
preconditionFailure("A body response continuation from this iterator already exists! Queuing calls to `next()` is not supported.")
```